### PR TITLE
Feat: add cellpose-inspired right-click to encircle mask creation interaction mode

### LINKED ISSUE #8363: Ability to auto-fill closed labels when painting
#5006 was a nearly-ready PR to close labels while painting, which can dramatically speed up annotation with Labels layers. (Like Lasso in Shapes but for Labels.) We should pick it up and make sure we credit the original contributor. 🙏 

> This PR adds a mode (tentatively) named 'fill_contour' to automatically fill a contour with currently selected label value once the drawn contour is closed. In other words, it 'auto-fills' a closed contour. When labeling (2D) images, having to press a mouse button down while dragging gets cumbersome when labeling many objects. While there had been a recent PR (easy switching to 'fill' mode via a keybinding) to alleviate some of this #4696, a *one handed mode of operations for labeling feels more natural and less strenuous*. In addition, while annotating objects of different sizes, one often needs to zoom in/out interactively much like the current 'pan_zoom', so this PR keeps this interactive mode while still allowing do draw without needing to press a key to toggle 'pan_zoom' mode.
> 
> To start drawing a contour with the currently selected label, a right-click will toggle the drawing mode ON (the left mouse click is used to drag/move the image as in 'pan_zoom' mode). Once the contour is closed (the last point being drawn has returned to the initial location), the contour is filled with currently selected label, and the selected label is automatically incremented to the next available number. 
> 
> If a closed contour is not desired, a second right-click will toggle the drawing mode OFF. 
> 
> While in 'fill_contour' mode, a circular brush size is used and its size is updated to reflect the current zoom.
> 
> The behavior of 'fill_contour' mode assumes that once a label is created, another label should not draw over the existing labels. This is to prevent accidentally drawing over other labels when annotating a crowded field where objects are touching.
> 
> *Note: I've quickly drawn a little icon to go with this mode, but it's currently just a placeholder. It would be cool to design something that makes more sense for the mode.*
> 
> napari_fill-contour_mode_01

_Originally posted by @delnatan in https://github.com/napari/napari/issues/5006

### PR DESCRIPTION
**Pull request by tlambert03**
# Description

Was chatting with @carsen-stringer this week and she mentioned a mouse interaction feature from cellpose that she's long requested/wanted in napari. This PR implements it (explained below).  Since it adds a new interaction mode to an already "full" set of interaction modes, i understand you'll want to consider it carefully.  There's no presumption that this will get merged easily, I just figured it would be easier to discuss with a working PR.  Apologies if this has been previously proposed and rejected. 

*edit: claude searched git history and found a few related issues, so this has some history*:
-  #5006
-  closes #8363

### The core feature

Cellpose has, I would argue, a "higher throughput" lasso/encircle interaction mode than the available interactions in napari, that allows one to more quickly generate multiple masks.  You (right)-click (and release), then move your mouse around the perimeter of the object, and when you return to within some radius of the original click point (marked by a circle) the label is completed and filled in.  As implemented in this PR, it looks like this:

https://github.com/user-attachments/assets/cfff228b-657c-4e6d-a8ca-0999bc6fca92

### napari's current alternatives

Obviously, there are ways to accomplish similar things in napari already, but they are each slightly more laborious, and/or error prone.

- **you could use the polygon tool**.  This is the most similar to the goal of encircling a region.  The main difference is that the polygon requires _many_ mouse clicks along the way, and results in straight segmented edges.  You can of course use it very carefully with a *lot* of mouse clicks, and end up with something that is arguably even more precise than the new proposed mode (which locks in the outline immediately on mouse move), but it's a tradeoff between speed and non-straight edges.
- **you could use a two step pain brush + fill tool**.  Key combos make switching between the two tools pretty easy, but this requires switching to the paintbrush, *keeping* the mouse/trackpad pressed as you move the mouse (which is a bit hard on a trackpad), and then making a pixel-perfect closing of the loop (a single pixel gap would ruin the fill tool), then switching tools and clicking inside.

### Implementation details:

This PR overloads the paint tool.  Currently a right click behaves exactly like a left click (is that correct?).  With this PR, if you _right_ click while using the paint tool, it enters a temporary encircle mode where, as you move the mouse, it paints pixels as usual (though note that you don't have to keep the mouse pressed this time) until you either abort the motion using the escape key, or complete the motion by coming within a brush-size-determined radius of the original click point.

While cellpose auto-increments the label count each time, I chose not to implement that logic here.  User still needs to increment the label manually, using the `+` key or the GUI.

to try it out, clone, make labels layer, select paintbrush, and right click

---

I'll leave it to you all to discuss whether the new mode is different enough to warrant inclusion (no offense taken here if not). If you like parts of it, but not certain decisions, happy to change.  @carsen-stringer may have additional comments/pleas for considering its inclusion :)

- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to docstrings and documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
Love it! I know I've wanted something like the polygon lasso in Shapes for Labels. And autofill here is like https://github.com/napari/napari/pull/5006 
I think this would close https://github.com/napari/napari/issues/8363

**Comment by tlambert03** [comment]
> I think this would close https://github.com/napari/napari/issues/8363

yeah, i saw that after posting.  I added an edit above to the first comment.  i know this re-reraises a long discussion.  so, go back at it :joy: ... I don't have strong opinions here.  Just opening as a favor to @carsen-stringer

**Comment by psobolewskiPhD** [comment]
What controls the distance to autoclose? playing with cells 3d and the default results in clipped off slivers of nuclei because it close radius is large.

Edit: I'm very keen on using UX from other tools that people use. Less cognitive load if people are familiar with stuff!!

**Comment by tlambert03** [comment]
> What controls the distance to autoclose? playing with cells 3d and the default results in clipped off slivers of nuclei because it close radius is large.

it scales with brush radius.  The current proportionality constant is 1.8, which is what cellpose used... but I agree, it felt a bit too large in some cases (leading to a flat/clipped last segment).  Should be as easy as reducing that constant

**Comment by psobolewskiPhD** [comment]
Thanks -- sorry I didn't look at the code 😬 
Makes sense I was using a bigger brush.

One thing that keeps throwing me -- on laptop trackpad, macOS -- is with regular brush I use 3-fingers to paint, and with this i have to remember to use just 1, 3 fingers does pan -- which I guess is kinda cool to pan while painting...

Edit: what's wild is you can change the label while painting and it I guess works as expected? 🤔

**Comment by psobolewskiPhD** [comment]
What is pre-commit complaining about? data is defined like 4 lines higher, as the return from make_overlay...

**Comment by tlambert03** [comment]
yep, I'm equally confused there 🤷 ... i *had* been wrong in a previous commit. so maybe it's just not running the current sha?

**Comment by psobolewskiPhD** [comment]
Finally started looking at the code; it's very clever! I like how you use the existing polygon path for the close/fill.

File: src/napari/_vispy/overlays/labels_brush_stroke.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+    @_only_when_overlay_enabled
+    def _on_mouse_press(self, layer, event):
+        if not self.overlay.active:
+            # the right button starts an encircle-and-fill stroke
+            if event.button == 2:
+                self._start_stroke(layer, event)
+            return
+        if event.button == 2:
+            return
```
Comment (psobolewskiPhD) [inline src/napari/_vispy/overlays/labels_brush_stroke.py:106]: How about we also let the user complete by repeating the click? This would mimic the polygon lasso in shapes, which can be completed by distance (default is 0, but it's settable in Prefs) or by 2nd click.
```suggestion
        if event.button == 2:
            self._complete(layer)
            return
```

**Comment by psobolewskiPhD** [comment]
So I'm like 100% in favor of this, but thinking it maybe would be better as a tool/mode?
I just like this feature too much to have it not more easily discovered.
Advantage of a tool would be 1. discoverability 2. use of regular click instead of right click 3. better harmonization of annotation tools (similarity to polygon lasso).

We'd have two rows of buttons:
[shuffle] [pick] [pan/zoom] [transform]
[erase] [paint] [polygon] [lasso] [fill]
- re-use the icon from the shapes lasso

This is by no means a blocker, I'm just thinking out loud. 
I'd be happy with this as is and breaking it out into a mode as a followup. 
That said the implementation is so clean and clever I think making it a mode would not add much weight to the PR.
@jni would love your input on this, since you were involved in the prior art #5006 and the issue.

File: src/napari/_vispy/overlays/labels_brush_stroke.py (hunk 1 above)
Comment (carsen-stringer) [reply src/napari/_vispy/overlays/labels_brush_stroke.py:106]: Clicking at the end is additional mental energy for the annotator, it would be nice not to have to do this, or have that option if possible

Comment (psobolewskiPhD) [reply src/napari/_vispy/overlays/labels_brush_stroke.py:106]: With that change both work. Either draw to the end, or click to end early. It mimics the behavior of our Shapes lasso tool.

Comment (psobolewskiPhD) [reply src/napari/_vispy/overlays/labels_brush_stroke.py:106]: I was doing some annotations of things that were not circular and the easy finish was a nice thing. Obviously it will depend on the use case.

**Comment by psobolewskiPhD** [comment]
> So I'm like 100% in favor of this, but thinking it maybe would be better as a tool/mode?

I did a bad thing and had Deepseek just implement this (and change nomenclature from brush_stroke to lasso) while prepping diner.
I like it a lot...

Pushed it up to my fork if anyone wants to try it:
https://github.com/psobolewskiPhD/napari/tree/label_lasso

**Comment by brisvag** [comment]
Hi @tlambert03 :wave: My first impression is that the thing we discussed in the past (sharing the same logic and code path with the shape lasso tool via an overlay, and then completing the action differently depending on the layer) would be a bit cleaner. I don't really see the benefit of doing thhis action with a brush, as that just makes it less precise the more the brush is wide.

**Comment by tlambert03** [comment]
thanks @brisvag,  can you be a bit more specific? (sorry i wasn't part of that discussion, you can link a comment if you're looking at one).  Is what you're describing basically the "**you could use the polygon tool.**" section of my original post?  something where the user has to click control points continuously as they make their way around the object?

> I don't really see the benefit of doing thhis action with a brush, as that just makes it less precise the more the brush is wide.

If i'm not mistaken, this _is_ the specific feature request that some (like @carsen-stringer) would want.   (You can of course, make the brush smaller if you want precision, just as in cellpose).  But if napari core devs are specifically rejecting the interaction mode of instant circling (i.e. as you move the mouse, it's immediately charting the outline, rather than a more precise, but slower, click-per-control-point) then that is indeed probably the decision that would just close this feature as "not planned"?

**Comment by brisvag** [comment]
> Is what you're describing basically the "**you could use the polygon tool.**" section of my original post?

Nope, `polygon lasso` is a different mode from `polygon` on the shapes layer (I argued to merge the two at the time, I think the naming/splitting is confusing :P). This mode still uses a path-based outline (so not the brush), but handles mouse input smoothly (like your method does) without having to click. It even automatically serves 2 different modes: quick-click-release and then mouse move (nice cause it doesn't force you to hold the click), versus click-hold-drag, which is better for pen/pad users. I think that tool covers this same use case and it would be best to unify that to the current polygon tool of the labels layer.

> If i'm not mistaken, this _is_ the specific feature request that some (like @carsen-stringer) would want. (You can of course, make the brush smaller if you want precision, just as in cellpose).

Sure, if folks want this let's see how to do this instead. I'm just asking "why", since I don't really see a benefit of having the thick brush in this specific use case :shrug: It sound more to me that the actual need is for a more ergonomic mouse input, and not specifically for the brush to have a thickness when doing the circling? In fact, I'm ready to bet that the thickness is most of the time a negative since you have to move **beside** the edge insted of **on** it, but that's just how cellpose happens to have done it...

**Comment by tlambert03** [comment]
> Nope, polygon lasso is a different mode from polygon on the shapes layer 

ahhh... you're talking about a mode from the _shapes_ layer.  So, you're proposing reusing a mechanism that already exists on shapes, and just bringing it to labels.  That certainly makes sense.  I can look into it.  

Regarding "why":  Yes, you're absolutely correct I don't think anyone particularly cares specifically about retaining the _thickness_ of the brush.  It _is_ just about a more ergonomic mouse movement.  Totally agree that thickness is awkward, and I have been moving it to a brush of 1-2 when using this.  I used brush because I saw some previous discussion that seemed to have stalled prior attempts because new modes were added to the already full UI buttons, and since the behavior (instant commit without click) was more brush-like, I attached it to the brush.  

@carsen-stringer, can you confirm that you don't really care about the thickness of the brush for this feature?  Is there ever a time when you actually prefer to use a larger brush when encircling a region?

**Comment by carsen-stringer** [comment]
Sorry for @brisvag for not knowing about that mode, yes the brush thickness doesn't matter. If you were to do it with the brush, I would have some opacity on, that's what we do in Cellpose, so you can keep the brush approximately centered on the edge of the cell. The neural networks are pretty robust to small mistakes, so we just tried to implement what we thought was the fastest for 2D labels. Also in Cellpose, the closing region area (the red circle) is proportional to the brush size - this is one thing that users might want to have control over.

I just want to add that Napari is amazing!! We started building the Cellpose GUI before Napari was released, and it is very much focused on our use-case. If we were able to label quickly in Napari, then we could move our model fine-tuning functionality to a Napari plugin (which is fairly widely used). But also no pressure from us. And thank you @tlambert03 for implementing this! And @psobolewskiPhD for extending it / testing it

**Comment by psobolewskiPhD** [comment]
I think we absolutely should get some variant of this PR in. It's a feature that's been in demand for a long time!

The only question for me is whether it should be a brush tool variant as now, using the right click, or it's own tool/mode (e.g. as I prototyped here https://github.com/psobolewskiPhD/napari/tree/label_lasso ). 

I lean to the mode/tool because it's easier to discover and we can more easily gate other behavior on that, e.g. implementing auto-increment becomes trivial for a mode/tool (this new one plus the existing polygon tool). In that case copying at least the semantics of the Shapes polygon lasso makes sense IMO (e.g. use regular click, allow click to close early). 

Copying the full behavior to me is less important, because the layers do behavior quite differently. For example, I find being able to make a large brush useful for annotating large areas e.g. tissue slices in a WSI, because you don't need to be so precise, you can let the curvature of the brush take care of things. Note that the current implementation also lets you drop the brush size down if you want.

(Note: i like the circle indicator for the close-zone, we should port that to the shapes lasso tool which has the same feature -- user controlled in the settings -- but no indicator)

**Comment by brisvag** [comment]
> For example, I find being able to make a large brush useful for annotating large areas e.g. tissue slices in a WSI, because you don't need to be so precise, you can let the curvature of the brush take care of things.

Good point, this does fit better into the typical interaction with the labels layer.

**Review by jni — APPROVED** [review]
Thank you @tlambert03!

I think there's a lot of ways to get this mode in, and therefore lots of potential bikeshedding. In my opinion, right-click is very low risk since it is presently unused. Every other option discussed is more intrusive (involving rearranging the button layout), so I think we should merge, get it in people's hands, and then iterate on the UX.

**Comment by tlambert03** [comment]
Thanks @jni 

I agree, there's lots of reasonable ways this can be done, and that's probably why it still hasn't been done :) 
I'm of course happy to make any changes if the core devs come to a consensus.  Just let me know the final bullet points of what the discussion yields (if it happens) and I'm happy to do the implementation.

**Comment by TimMonko** [comment]
@tlambert03, I must confess that I've been meaning to review this for 3 weeks now. Yes, it's partly because I have some other things I need to prioritize, but also because I got into thinking much too into the weeds.

_Everyone_ I've spoken to absolutely loves this: Grzegorz, Carlos, Draga, Sesan, Jacopo.

@jni  encouraged us yesterday to not worry so much about the exact way to implement this -- and work on shipping it because, indeed, it is truly awesome. 
I'll use this opportunity as a good example going forward on how we can better support these big features, and I just wanted to thank you for your time and patience with us as we sorted ourselves out.

I played with this first like 3 weeks ago, and was too caught up in trying to even sort out how we could potentially expose it (new button? right click? A settable right click to any mode?... etc) and how to move it forward. 
Yet, when I first played with it, it worked great. I'm glad that Juan helped us move forward and merge it. I look forward to if/how this gets iterated on going forward! 

Certainly better is the enemy of good -- but you've already got great here! So thank you for taking the time to make this great feature!

**Comment by jni** [comment]
I made #9193 for follow-up discussions.
